pkg/reconciler/apis/apiexport/apiexport_controller.go

```diff
@@ -0,0 +1,489 @@
+				for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
+					schemaKey := clusters.ToClusterAwareKey(logicalcluster.From(apiExport), schemaName)
+					schema, err := apiResourceSchemaInformer.Lister().Get(schemaKey)
+					if err != nil {
+						// TODO log
+						continue
+					}
+
+					ret = append(ret, fmt.Sprintf("%s/%s/%s", apiExport.Status.IdentityHash, schema.Spec.Group, schema.Spec.Names.Plural))
+				}
+				return ret, nil
+			},
+		},
+	); err != nil {
+		return nil, err
+	}
+
+	apiExportInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIExport(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+	})
+
+	secretInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueSecret(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueSecret(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueSecret(obj)
+		},
+	})
+
+	return c, nil
+}
+
+// controller reconciles APIExports. It ensures an export's identity secret exists and is valid.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClusterClient kcpclient.ClusterInterface
+	apiExportLister  apislisters.APIExportLister
+
+	kubeClusterClient kubernetes.ClusterInterface
+
+	getNamespace    func(clusterName logicalcluster.LogicalCluster, name string) (*corev1.Namespace, error)
+	createNamespace func(ctx context.Context, clusterName logicalcluster.LogicalCluster, ns *corev1.Namespace) error
+
+	secretLister    corelisters.SecretLister
+	secretNamespace string
+
+	getSecret    func(ctx context.Context, clusterName logicalcluster.LogicalCluster, ns, name string) (*corev1.Secret, error)
+	createSecret func(ctx context.Context, clusterName logicalcluster.LogicalCluster, secret *corev1.Secret) error
+}
+
+// enqueueAPIBinding enqueues an APIExport .
+func (c *controller) enqueueAPIExport(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.V(4).Infof("Queueing APIExport %q", key)
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueSecret(obj interface{}) {
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("unexpected object type: %T", obj))
+			return
+		}
+
+		secret, ok = tombstone.Obj.(*corev1.Secret)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("unexpected object type: %T", obj))
+			return
+		}
+	}
+
+	if secret.Annotations == nil {
+		return
+	}
+
+	apiExportCluster := secret.Annotations[apisv1alpha1.AnnotationKeyAPIExportCluster]
+	apiExportName := secret.Annotations[apisv1alpha1.AnnotationKeyAPIExportName]
+	if apiExportCluster == "" || apiExportName == "" {
+		return
+	}
```

Comment:
I think this coupling of the secret and the export is not good. We should make the secret reusable, and also very important: never delete it, i.e. not owner reference (haven't checked whether we have set one).

In other words, here we have to go through all exports (there won't be many, usually one) and check the secret ref.

Reply:
Fixed